### PR TITLE
[python] Drop Python 2 support

### PR DESCRIPTION
**Pull request by StrikerRUS**
Do not include in any **patch** `3.1.x` release. Should be merged only in `3.2.0` version.

### CODE REVIEW

File: docker/dockerfile-r
Comment (jameslamb) [inline docker/dockerfile-r:9]: I support this change, but it seems unrelated. Could you put it in a separate PR that we could approve and merge more quickly (since this PR will have to wait for a while until we know if the next release will be 3.1.1 or 3.2.0)?

File: tests/python_package_test/utils.py
Comment (jameslamb) [inline tests/python_package_test/utils.py:20]: I'm very happy to see this go away :grinning:

**Review by jameslamb — CHANGES_REQUESTED** [review]
These changes look great to me, thanks! I left one suggestion for an unrelated change to be moved.

File: docker/dockerfile-r
Comment (StrikerRUS) [reply docker/dockerfile-r:9]: OK, sure!

**Review by jameslamb — APPROVED** [review]
Really well-done! I think you got everything that's dependent on Python 2. I can't think of anything else.

Tagging @aldanor @synapticarbors as maintainers of https://github.com/conda-forge/lightgbm-feedstock, to let them know that the next LightGBM minor release (3.2.0) will drop Python 2 support.

**Comment by StrikerRUS** [comment]
@jameslamb 
> I can't think of anything else.

One more thing is string formatting unification (`str.format()` at least or cute f-strings). But I think it is a great candidate for Hacktoberfest 2021. 🙂

File: python-package/lightgbm/compat.py
Comment (aldanor) [inline python-package/lightgbm/compat.py:168]: Could there be some user code downstream catching/filtering this one?

Comment (StrikerRUS) [reply python-package/lightgbm/compat.py:168]: Sorry, could you elaborate on what code you'd expect?

Comment (aldanor) [reply python-package/lightgbm/compat.py:168]: It's probably ok, I don't think anything needs to be done :) What I meant is that you could say it's *technically* a breaking change since there can be user code like below:

```python
from lightgbm.compat import LGBBDeprecationWarning
import warnings
warnings.simplefilter('ignore', LGBMDeprecationWarning)
```

Comment (StrikerRUS) [reply python-package/lightgbm/compat.py:168]: Ah, OK, got it!

Yeah, after this PR is merged, one should change 
```
from lightgbm.compat import LGBMDeprecationWarning
```
to 

```
from lightgbm.basic import LGBMDeprecationWarning
```

But we are planing to include this PR in `3.2.0` (not `3.1.1` patch) version and it has `breaking` label in the changelog. So, I believe it is OK...

**Comment by StrikerRUS** [comment]
@guolinke Are you OK with this PR?

**Review by guolinke — APPROVED** [review]
LGTM

**Pull request merged**
